src/app/core/export/query.service.ts

```diff
@@ -82,33 +99,51 @@ export class QueryService {
     }).value;
   }
 
-  private async loadData(from: Date, to: Date): Promise<void> {
-    const entityClasses: [EntityConstructor<any>, () => Promise<Entity[]>][] = [
-      [Child, () => this.entityMapper.loadType(Child)],
-      [School, () => this.entityMapper.loadType(School)],
-      [RecurringActivity, () => this.entityMapper.loadType(RecurringActivity)],
-      [
-        ChildSchoolRelation,
-        () => this.entityMapper.loadType(ChildSchoolRelation),
-      ],
-      [Note, () => this.childrenService.getNotesInTimespan(from, to)],
-      [EventNote, () => this.attendanceService.getEventsOnDate(from, to)],
-    ];
-
-    this.entities = {};
-
-    const dataPromises = [];
-    for (const entityLoader of entityClasses) {
-      const promise = entityLoader[1]().then((loadedEntities) =>
-        this.setEntities(entityLoader[0], loadedEntities)
-      );
-      dataPromises.push(promise);
-    }
+  private cacheRequiredData(query: string, from: Date, to: Date) {
+    const uncachedEntities = this.getUncachedEntities(query, from, to);
+    const dataPromises = uncachedEntities.map((entity) => {
+      if (this.dataFunctions[entity.ENTITY_TYPE]) {
+        return this.dataFunctions[entity.ENTITY_TYPE](from, to).then(
+          (loadedEntities) => {
+            this.setEntities(entity, loadedEntities);
+            this.dataLoaded[entity.ENTITY_TYPE] = { from, to };
+          }
+        );
+      } else {
+        return this.entityMapper.loadType(entity).then((loadedEntities) => {
+          this.setEntities(entity, loadedEntities);
+          this.dataLoaded[entity.ENTITY_TYPE] = true;
+        });
+      }
+    });
 
-    await Promise.all(dataPromises);
+    return Promise.all(dataPromises);
+  }
 
-    this.dataAvailableFrom = from;
-    this.dataAvailableTo = to;
+  /**
+   * Get entities that are referenced in the query string and are not sufficiently cached.
+   * @param query
+   * @param from
+   * @param to
+   * @private
+   */
+  private getUncachedEntities(query: string, from: Date, to: Date) {
+    return this.queryStringMap
+      .filter(([matcher]) =>
+        // matches query string without any alphanumeric characters before or after (e.g. so Child does not match ChildSchoolRelation)
+        query.match(new RegExp(`(^|\\W)${matcher}(\\W|$)`))
+      )
+      .map(([_, entity]) => entity)
+      .filter((entity) => {
+        const loadingInfo = this.dataLoaded[entity.ENTITY_TYPE];
+        return (
+          loadingInfo === undefined ||
+          !(
+            loadingInfo === true ||
+            (loadingInfo.from <= from && loadingInfo.to >= to)
+          )
+        );
+      });
```

Comment:
this is a slightly complex pyramid of maps for keys and cache status. Still reasonably easy to understand but possibly there could be a way to have a single data construct to keep it all? (e.g. query regex, loading function, cached status, ... as part of one object?) Not sure whether this would make things better though, just a random thought.